Core/src/org/sleuthkit/autopsy/datamodel/BlackboardArtifactNode.java

```diff
@@ -1040,6 +955,196 @@ public String getItemType() {
+         */
+        @Beta
+        static public BlackboardArtifactNodeKey createNodeKey(BlackboardArtifact artifact, boolean lookupIsAssociatedFile) throws TskCoreException {
+            try {
+                return keyCache.get(artifact.getObjectID(), () -> new BlackboardArtifactNodeKey(artifact, lookupIsAssociatedFile));
+            } catch (ExecutionException ex) {
+                throw new TskCoreException(String.format("Failed to get node key for artifact from cache id(%d)", artifact.getId()), ex);
+            }
+        }
+
+        /**
+         * Creates a new key for the given object.
+         *
+         * Note: This method is public to support subclass in CVT.
+         *
+         * @param artifact
+         */
+        public BlackboardArtifactNodeKey(BlackboardArtifact artifact) {
+            this(artifact, false);
+        }
+
+        /**
+         * To work with the "beta" constructor that uses the associated file as
+         * the source content.
+         *
+         * @param artifact
+         * @param lookupIsAssociatedFile
+         */
+        private BlackboardArtifactNodeKey(BlackboardArtifact artifact, boolean lookupIsAssociatedFile) {
+            this.artifact = artifact;
+            this.lookupIsAssociatedFile = lookupIsAssociatedFile;
+            try {
+                initializeKeyData();
+            } catch (TskCoreException ex) {
+                logger.log(Level.SEVERE, MessageFormat.format("Failed to create artifact node key for (artifact objID={0})", artifact.getId()), ex); //NON-NLS
+            }
+        }
+
+        /**
+         * Initializes the data for the key.
+         *
+         * @throws TskCoreException
+         */
+        private void initializeKeyData() throws TskCoreException {
+            if (artifact == null) {
+                return;
+            }
+
+            if (lookupIsAssociatedFile) {
+                try {
+                    associatedFile = getPathIdFile(artifact);
+                } catch (ExecutionException ex) {
+                    logger.log(Level.SEVERE, String.format("Failed to getPathIdFile for artifact (%d)", artifact.getObjectID()), ex);
+                }
+            }
+
+            final long objectID = artifact.getObjectID();
+            try {
+                sourceContent = contentCache.get(objectID, () -> artifact.getSleuthkitCase().getContentById(objectID));
+            } catch (ExecutionException ex) {
+                throw new TskCoreException("", ex);
+            }
+
+            // These calls cache their values in the artifact\content objects.
+            artifact.getAttributes();
+            sourceContent.getParent();
+            sourceContent.getUniquePath();
+
+            Content dataSource = sourceContent.getDataSource();
+            if (dataSource != null) {
+                dataSourceStr = dataSource.getName();
+            } else {
+                dataSourceStr = getRootAncestorName(sourceContent);
+            }
```

Comment:
@rcordovano 
_we are throwing away the data source object we have constructed after harvesting its name. Are we trying to conserve memory here?_
I took a look at AbstractContent.getDataSource 

`public Content getDataSource() throws TskCoreException {
		Content myParent = getParent();
		if (myParent == null) {
			return null;
		}

		return myParent.getDataSource();
	}`

AbstractContent.getParent caches the parent object after a lazy construction.

`public Content getParent() throws TskCoreException {
		// It is possible that multiple threads could be doing this calculation
		// simultaneously, but it's worth the potential extra processing to prevent deadlocks.
		if (parent == null) {
			ObjectInfo parentInfo;
			parentInfo = db.getParentInfo(this);
			if (parentInfo == null) {
				parent = null;
			} else {
				parent = db.getContentById(parentInfo.getId());
			}
		}
		return parent;
	}`

I didn't see the need to store the dataSource object and I don't think we need to be concerned if it is called in other places.  Am I incorrect in my interpretation?

Reply:
@rcordovano I realize the flaw in my above comment... not all Content objects are also AbstractContent.  BlackboardArtifact does not subclass AbstractContent.

Reply:
The implementation of getDataSource() in AbstractContent looks a little weird to me. Every TSK object should have a data source. Even data sources have data sources - themselves. In Image:
    
`	@Override
	public Content getDataSource() {
		return this;
	}`

and in LocalFilesDataSource:

`	@Override
	public Content getDataSource() throws TskCoreException {
		if (this.getDataSourceObjectId() == this.getId()) {
			// This virtual directory is a data source.
			return this;
		} else {
			return super.getDataSource();
		}
	}`

The getDataSource() override in BlackboardArtifact should just look up the data source using the dataSourceObjId field. Please write a story about the implementation in AbstractContent, thanks!